### PR TITLE
feat: add CZI and ND2 support to cell detection

### PR DESCRIPTION
**Pull request by adiezsanchez**
Skill now can process raw data from Zeiss .czi and Nikon .nd2 files.

Multichannel 3D stacks are supported by reducing the Z axis by max projection.

- add CZI and ND2 support to cell detection
- add tests for CZI and ND2 support
- update requirements.txt to include czifile and nd2
- update SKILL.md to include CZI and ND2 support
- update cell_detection.py to include CZI and ND2 support
- update tests/test_cell_detection.py to include CZI and ND2 support
- update README.md to include CZI and ND2 support
- update README.md to include CZI and ND2 support

## Summary

- Add direct Zeiss `.czi` and Nikon `.nd2` support to `cell-detection` image loading.
- Support multichannel 3D stacks by reducing the Z axis using max projection before segmentation.
- Update tests, skill docs, and dependencies to cover and document the new microscopy input formats.
- Get in touch for demo data (files too big for commit)

## Skill affected

`cell-detection`

## Checklist

- [x] SKILL.md is present and complete (YAML frontmatter + methodology)
- [x] Tests included and passing (`pytest -v`)
- [] Demo data included for reviewers to test
- [x] No patient/sensitive data committed
- [x] Follows [CONTRIBUTING.md](../CONTRIBUTING.md) conventions

## Test output

```bash
python -m pytest skills/cell-detection/tests/test_cell_detection.py -k "czi or nd2" -q
..                                                                       [100%]
2 passed, 64 deselected, 1 warning in 1.01s
```

```bash
python -m pytest skills/cell-detection/tests/test_cell_detection.py -k "TestLoadImage" -q
.....                                                                    [100%]
5 passed, 61 deselected, 1 warning in 1.33s
```

### CODE REVIEW

**Comment by camlloyd** [comment]
Hi @adiezsanchez, thanks for this! What do you think about using something like [bioio-devs/bioio](https://github.com/bioio-devs/bioio) readers here?

**Comment by adiezsanchez** [comment]
Hei @camlloyd my inclination is to keep direct support for formats like .czi and .nd2 using their native libraries, since BioIO can be a bit cumbersome for users due to the extra plugins required. On the other hand, BioIO does provide a cleaner abstraction layer and might make it easier to support additional formats in the future. Maybe in a next iteration of the skill along with multichannel 3D stack cell counting and segmentation?

**Review by manuelcorpas — COMMENTED** [review]
**Audit verdict: needs review**

Thanks for adding CZI and ND2 support. Dependencies (`czifile`, `nd2`, Pillow) are legitimate, no typosquats. Two things before merge:

1. 4D axis handling (`cell_detection.py` ~lines 89-98): the channel-vs-Z disambiguation is inferred from "whichever axis is smaller" rather than the format's known layout (CZI is C-Z-H-W, ND2 is Z-C-H-W). On a common low-plane confocal stack (e.g. 3 channels x 2 planes) this picks the wrong axis and max-projects across fluorescence channels, silently corrupting detection with no error. The format is known at read time and both readers expose authoritative axis labels (`CziFile.axes`, nd2 `.sizes`); please use them instead of guessing, rather than discarding them and calling `.squeeze()`.
2. Tests do not exercise the failing path: the CZI fixture `(1,3,32,32)` squeezes to 3D so the 4D block never runs, and the ND2 fixture is a single happy-path ordering. Please add a C-Z vs Z-C divergence test and a bit-depth/scaling assertion (CZI/ND2 are commonly uint16).

Note (not blocking): `nd2` parses files via a vendored native Nikon SDK, so untrusted ND2 input is a native-code surface; fine for trusted data.


---
*Automated audit by ClawBio pr-audit. Manual review is still recommended.*

**Review by manuelcorpas — COMMENTED** [review]
## Audit verdict: NEEDS REVIEW (correctness, not security)

No security, supply-chain or data-movement risk: `czifile` (Gohlke) and `nd2` (Lambert) are the canonical readers, not typosquats; the loader only takes a local path; no `np.random` in the reported results. The issue is axis handling.

**1. Channel-vs-z axis is guessed from array shape, not read from format metadata (`cell_detection.py:86-98`).** Both readers expose explicit axis labels (czifile `CziFile.axes`, e.g. `STCZYX`; nd2 `ND2File.sizes` / xarray dims), but the code infers channel vs z by "whichever small axis is smaller". A stack with more channels than z-planes (e.g. `(4,3,H,W)`) gets the wrong axis collapsed and channels scrambled. cpsam is channel-order invariant, so this never throws, it just produces plausible-but-wrong masks. Please drive the axis selection from the metadata both readers expose.

**2. Unconditional `.squeeze()` on CZI (`cell_detection.py:72-75`)** destroys a singleton channel axis: a 1-channel/8-z CZI squeezes to `(8,H,W)` and the 8 z-planes are read as 8 channels (then truncated to 3). The ND2 path does not squeeze, so the two formats are handled inconsistently.

**3. The new tests are mocked around the failing path (`tests/test_cell_detection.py:344-365`).** The CZI fixture `(1,3,32,32)` squeezes to 3D and never enters the 4D branch; the ND2 fixture happens to be the one orientation the heuristic gets right. Please add real-shape fixtures covering channel>z and singleton-channel cases.

Minor: pixel-size/µm calibration from CZI/ND2 is discarded (measurements stay in pixels, internally consistent but not physically comparable across magnifications), and `czifile`/`nd2` are added unpinned. Worth a floor pin and a note in the report.

Automated audit by ClawBio pr-audit. Manual review is still recommended.

**Review by manuelcorpas — COMMENTED** [review]
**ClawBio domain audit — NEEDS REVIEW (correctness)**

No security or clinical risk, no regression to the existing TIFF / PNG / JPG paths, the new deps (`czifile`, `nd2`) are declared consistently across requirements.txt, SKILL.md and the environment.yml writer, and imports are lazily guarded. The concern is multi-dimensional axis handling:

- `cell_detection.py:111` does a blind `np.asarray(arr).squeeze()` and `:112` hardcodes `axes = ""`, discarding the file's own axis metadata. czifile exposes `CziFile(path).axes` and nd2 exposes named dims; axis assignment should be driven by those, not by a magnitude heuristic.
- `:122-134` disambiguates channel vs Z by size (<=20). For an ordinary 4-channel x 3-Z ND2 (`4x3xHxW`) it picks the smaller axis as channels, max-projecting across the 4 channels and treating the 3 Z-planes as channels: a silent wrong-axis projection.
- The Z max-projection (`:124,134`) is an irreversible default with no opt-out; please expose and document it.
- The two added tests use `np.ones` mocks whose shapes happen to dodge the failure mode (`test_cell_detection.py:347` is `4x3`, which resolves correctly by luck). Add a test with an ambiguous / realistic axis order that would fail under the current heuristic.

Requested: assign C / Z / H / W from the readers' axis labels, add the failing-case test, and document the Z projection as a deliberate, overridable choice.

Automated audit by ClawBio pr-audit. Manual review is still recommended.

**Comment by adiezsanchez** [comment]
@manuelcorpas and @camlloyd I’ve pushed changes to skills/cell-detection/ addressing the earlier audit findings. Summary for re-review:

**Blocking items addressed**

- CZI/ND2 C/Z mapping now uses CziFile.axes and ND2File.sizes (not shape heuristics alone)
- Removed blind .squeeze(); only non-C/Z/Y/X singleton dims are dropped
- Added regression tests for metadata paths, singleton channels, z_projection=none → Z×C×Y×X, heuristic 4D fallback, and 3D multichannel segmentation axes
- --z_projection max|none is documented and wired through CLI/report; 3D volumetric mode uses none + --do_3D

**Minor items addressed**

- Pinned czifile>=2019.7.2.2 and nd2>=0.11.1 (manually verified)
- Added TestCliGuardrails for --do_3D / --z_projection consistency
- Renamed misleading CZI test (test_load_czi_zcyx_metadata_max_projects_to_hwc)
- SKILL.md notes on pixel calibration and ND2 native-code surface

**Manual validation** 

- Tested on real files in clawbio_test/clawbio_test_data (cell_3D_3ch.nd2, organoid_3D_4ch.czi, organoid_2D_4ch.czi) — loads match expected shapes for both max and none projection modes. Unit tests: 78/79 pass in the clawbio conda env (one Windows-only commands.sh executable-bit failure).

Happy to walk through any part of the diff, re-run checks on your side. Let me know what is the easiest way to share the .czi and .nd2 files so you can check functionality on your side.

**Review by manuelcorpas — COMMENTED** [review]
## Audit verdict: NEEDS REVIEW

Deep domain re-audit. No malicious code, no remote data exfiltration, no fabricated results. Clearing the automated flags and noting one real bug.

### Automated CRITICALs are false positives - please disregard
The scanner flagged three `eval()` calls as CRITICAL. None is Python's builtin `eval`:
- `cell_detection.py:320` `model.eval(img, **eval_kwargs)` is cellpose's `CellposeModel.eval()` segmentation inference method.
- `tests/test_cell_detection.py` `def eval(self, ...)` is a method on the `_FakeCellposeModel` mock.
- The SKILL.md reference is to the same cellpose `.eval` API.

The "missing SKILL.md sections" flags are also false: the file is complete (174 lines), with Safety + ClawBio disclaimer, numbered Workflow, Input Formats, Output Structure, etc.

### Real finding (should fix before merge)
**Silent channel/Z mis-detection on metadata-less 4D files.** `cell_detection.py:262-272` guesses Z vs C by axis size when vendor metadata is absent or unrecognised:
```python
if arr.shape[0] <= 20 and arr.shape[1] <= 20:
    channel_axis = 0 if arr.shape[0] <= arr.shape[1] else 1
...
z_axis = 1 if channel_axis == 0 else 0
arr = arr.max(axis=z_axis)
```
For a real `Z x C x Y x X` stack where `Z <= C` (e.g. 3 z-slices x 4 channels) this max-projects over the true channel axis and feeds the z-planes to cellpose as channels, reporting `n_channels=3`. The metadata-driven happy path (`CziFile.axes` / `ND2File.sizes`) is correct and well-tested; the failing heuristic branch has no test (all CZI/ND2 tests mock readers with clean metadata, and the one "heuristic" test gets the right answer only because `Z > C`).

Suggested: hard-error on the ambiguous `Z <= 20 and C <= 20` case (or require an explicit `--channel_axis` / Z-axis flag), and add a test driving a metadata-less `Z <= C` 4D stack through `load_image`.

Minor: pin `Pillow` in `requirements.txt` (currently unpinned). `czifile` and `nd2` are legitimate, correctly attributed packages; local-first preserved.

---
*Automated audit by ClawBio pr-audit. Manual review is still recommended.*

**Comment by adiezsanchez** [comment]
@manuelcorpas Fixed, went for the hard-error route rather than adding even more CLI arguments, trying to keep workflow as automated as possible with minimal user input.

**Review by manuelcorpas — APPROVED** [review]
## Re-audit verdict: SAFE — prior finding fixed

Verified against the live head code and by running the test suite.

- **Silent Z/channel mis-detection fixed.** The metadata-less ambiguous 4D path (`Z <= 20 and C <= 20`, `Z <= C`) now raises `ValueError("Cannot infer channel vs Z axis...")` instead of silently max-projecting over a guessed axis. The metadata-driven path was refactored to run first and resolve C/Z deterministically.
- **Regression test added** (`test_load_4d_heuristic_raises_when_z_le_c_and_no_metadata`) covering the exact changed branch in both projection modes. Suite runs 76 passed / 4 failed, the 4 failures being environmental only (`cellpose` not installed locally; present in CI). The metadata happy path, including the C > Z case, is intact and tested.
- **Pillow pinned** in `requirements.txt`.
- The scanner's 3 `eval()` CRITICALs are false positives (cellpose `model.eval()` inference + a mock `def eval`), as before.

Two minor follow-ups (non-blocking): the SKILL.md install block still lists Pillow unpinned (drift vs `requirements.txt`), and `--use_gpu` / `--use_cpu` are not mutually exclusive at argparse level (CPU safely wins). Approving.

---
*Automated audit by ClawBio pr-audit. Manual review is still recommended.*
